frontend/admin/resources/js/components/AuthContainer.tsx

```diff
@@ -4,90 +4,159 @@ import {
   parseUrlQueryParameters,
   useLocation,
 } from "@common/helpers/router";
-import { homePath } from "../adminRoutes";
+import { homePath, refreshTokenPath } from "../adminRoutes";
 
 const ACCESS_TOKEN = "access_token";
 const REFRESH_TOKEN = "refresh_token";
+const TOKEN_EXPIRY = "token_expiry";
 
-interface AuthState {
+interface AuthContextState {
   loggedIn: boolean;
   accessToken: string | null;
   refreshToken: string | null;
+  expiry: number | null;
   logout: () => void;
+  refreshAuth: () => Promise<AuthState | null>;
+}
+
+interface AuthState {
+  accessToken: string | null;
+  refreshToken: string | null;
+  expiry: number | null;
 }
 
-export const AuthContext = React.createContext<AuthState>({
+export const AuthContext = React.createContext<AuthContextState>({
   loggedIn: false,
   accessToken: null,
   refreshToken: null,
+  expiry: null,
   logout: () => {
     /** do nothing */
   },
+  refreshAuth: () => Promise.resolve(null),
 });
 
-const logoutAndRefresh = (): void => {
+const logoutAndRefreshPage = (): void => {
   // To log out, remove tokens from local storage and do a hard refresh to clear anything cached by react.
   // TODO: Is there anything else we should do, in terms of notifying user?
   localStorage.removeItem(ACCESS_TOKEN);
   localStorage.removeItem(REFRESH_TOKEN);
+  localStorage.removeItem(TOKEN_EXPIRY);
   window.location.href = homePath();
 };
 
-function getTokensFromLocation(
+function calculateExpiryTimestamp(expiresIn: string | null): number | null {
+  const parsedExpiresIn = Number(expiresIn);
+  if (Number.isInteger(parsedExpiresIn)) {
+    return Date.now() + parsedExpiresIn * 1000; // expires_in is in seconds, not milliseconds
+  }
+  return null;
+}
+
+const refreshAuth = async (
+  refreshToken: string,
+  setAuthState: (state: AuthState) => void,
+): Promise<AuthState | null> => {
+  const response = await fetch(
+    `${refreshTokenPath()}?refresh_token=${refreshToken}`,
+  );
+  if (response.ok) {
+    const responseBody: {
+      access_token: string;
+      refresh_token: string;
+      expires_in: string | null;
+    } = await response.json();
+
+    const newAuthState: AuthState = {
+      accessToken: responseBody.access_token,
+      refreshToken: responseBody.refresh_token,
+      expiry: calculateExpiryTimestamp(responseBody.expires_in),
+    };
+
+    if (newAuthState.accessToken) {
+      setAuthState(newAuthState);
+      return newAuthState;
+    }
+  }
+  return null;
+};
+
+function getAuthFromLocation(
   location: ReturnType<typeof useLocation>,
-): { accessToken: string; refreshToken: string | null } | null {
+): AuthState | null {
   const queryParams = parseUrlQueryParameters(location);
   const accessToken: string | null = queryParams.access_token ?? null;
   const refreshToken: string | null = queryParams.refresh_token ?? null;
+  const expiry = calculateExpiryTimestamp(queryParams.expires_in);
```

Comment:
The access token itself contains its expiry time. Do we need to store it seperately?

Reply:
I decided to use the `expires_in` field since we get it for "free" and we don't need to add any other dependencies to get it.  You're right though, the expiry date is encoded in the token itself.  I'll try out jwt-decode though to see how much it adds to the bundle and how much it simplifies the code.

Reply:
Indeed, the size increase is very modest (about 16 KB) and it does simplify the code quite a bit.  Good call!